### DISCUSSION TITLE
How to detect object placed against a specific background?

### DISCUSSION THREAD

**Discussion by Yarr65** [ImageMagick/ImageMagick#2345] [Help] [answered]
I would like to check the correctness of transformation
of geographic point coordinates in a large number of graphic files.
I define an error as placing a point **completely** outside the cartographic outline.
What's the easiest way to detect files with an orange circle on a white background
as shown in the example below?

![valid](https://user-images.githubusercontent.com/69034223/89016284-c4a45000-d318-11ea-8acd-dc17eabc26f7.jpg)
**Valid**

![valid1](https://user-images.githubusercontent.com/69034223/89019713-04ba0180-d31e-11ea-9c69-172b8fa12e6c.jpg)
**Valid**

![not-valid](https://user-images.githubusercontent.com/69034223/89016280-c40bb980-d318-11ea-9e9e-bda0c843c70d.jpg)
**Not valid**

**Comment by snibgo** [comment]
What version of IM, on what platform?

The images are JPG, which is a lossy compression method. Sources that have not been lossy compressed would be better.

You show three examples. Example 1 has an orange circle entirely surrounded by brown, example 3 has an orange circle entirely surrounded by white, and example 2 has an orange circle surrounded by both brown and white.

These conditions can be detected. I don't know what makes your example 2 "valid".

How do we detect the conditions? Many approaches are possible. Here is one. It assumes the orange circle is _not_ at the edge of the image.

1. Remap the image to three colours: white, brown and orange. This removes all ambiguous anti-aliased colours.

2. Use connected-components to find the bounding-box of the orange circle. Crop out this rectangle, but with an extra pixel along each edge.

3. From that crop, make orange transparent, and turn brown into black. The mean of the result is either 0.0 (the orange circle was entirely surrounded by brown) or 1.0 (the orange circle was entirely surrounded by white) or between 0.0 and 1.0 (the orange circle was surrounded by both brown and white).

This algorithm is a bit complicated. A simpler method may be possible.

**Reply by Yarr65** [reply]
Excuse me. I should have prepared the problem better.
The second example is given solely to emphasize that the desired detection is not about catching the cases where the object touches white pixels, but only those where the object is completely surrounded by them.
Maybe let's forget the purpose of the application and simplify the matter as much as possible.

![v1](https://user-images.githubusercontent.com/69034223/89045992-ac4e2880-d34c-11ea-889a-e7e24f614610.png)**skip this case**
![v2](https://user-images.githubusercontent.com/69034223/89045991-ac4e2880-d34c-11ea-86de-d9ed78755f39.png)**skip this case also**

![error](https://user-images.githubusercontent.com/69034223/89045988-abb59200-d34c-11ea-9e12-add63748f378.png)**report only this case**

I just want a script that processes a directory of several thousand files to report to me:

 _Watch out buddy, your red dot is outside the green rectangle._

Of course, I will test your idea in practice (I hope it will not be too difficult for me).
Thank you.

**_Windows 10 Home | Intel Core i9-9000 CPU @ 3.1GHz 32GB RAM
ImageMagick 7.0.9-27 Q16 x64 2020-02-29_**

**Comment by snibgo** [comment, marked as answer]
Windows BAT script. Call it with the name of your red-dot file. Edit the stuff at the end to do whatever you want. All the "echo" commands are for debugging, and can be removed.
```
rem %1 is input file.

setlocal enabledelayedexpansion

set SRC=%1

rem input colours are #f00, #0f0 and #fff.

for /F "usebackq tokens=1-5" %%A in (`%IMG7%magick ^
  xc:#f00 xc:#0f0 xc:#fff +append -write mpr:MAP +delete ^
  %SRC% ^
  -remap mpr:MAP ^
  +write tmpinp.png ^
  -define "connected-components:verbose=true" ^
  -connected-components 4 ^
  NULL:`) do (
  echo %%A %%B %%C %%D %%E
  if "%%E"=="srgb(255,0,0)" (
    echo %%A %%B %%C %%D %%E
    for /F "tokens=1-4 delims=x+" %%G in ("%%B") do (
      echo %%G %%H %%I %%J
      set /A WW=%%G+2
      set /A HH=%%H+2
      set /A XX=%%I-1
      set /A YY=%%J-1
    )
  )
)

echo %WW% %HH% %XX% %YY%


for /F "usebackq" %%L in (`%IMG7%magick ^
  tmpinp.png ^
  -crop %WW%x%HH%+%XX%+%YY% ^
  +repage ^
  -fill Black -opaque #0f0 ^
  -transparent #f00 ^
  -scale "1x1^!" ^
  -alpha off ^
  -format "%%[fx:mean==0?1:mean<1?2:3]" ^
  info:`) do set CASE=%%L

echo CASE=%CASE%

if %CASE%==1 (
  echo case 1: red dot is entirely within green
) else if %CASE%==2 (
  echo case 2: red dot is on border of green and white
) else if %CASE%==3 (
  echo case 3: red dot is entirely outside green
) else (
  echo BUG: CASE=%CASE%
)
```

**Comment by Yarr65** [comment]
Thank you, **snibgo**.

Your script is very clear what helped me understand your idea.
I will certainly use it more than once.

I got a little excited and tried to simplify the procedure as much as possible.
So I paste my _scraps_ below:
```
set SRC=%1
set result=1
for /F "usebackq tokens=1-4" %%A in (`convert %SRC% ^
	-fill White +opaque "rgb(255,0,0)" ^
	-trim -format "%%[fx:w+4] %%[fx:h+4] %%[fx:page.x-2] %%[fx:page.y-2]" ^
	info:`) do (
	for /F "usebackq delims=(, tokens=3" %%E in (`convert %SRC% ^
		-crop %%Ax%%B+%%C+%%D +repage ^
		+dither -colors 4 -unique-colors ^
		txt:`) do if %%E LSS 255 set result=0
)
if !result!==1 echo %SRC%
```
I assume an error and only look to the contrary.

I use the fact that the **Red** value in the **RGB** table
of both desired colors (**Red** and **White**) is identical
and in the identification table I only look
for confirmation of the presence of a color
with an **R** value lower than **255** (these are the colors of the shape).

In the example from my workshop, the colors are a bit different,
but there also the **R** value of the circle color is higher
than the **R** value of the background pattern colors.
So I only operate on the lower threshold (**210**) with similar effectiveness.

The analysis of **2,800** files took a few minutes.
Only **20** of them required attention and only a few turned out to be incorrect.
A good result.
